File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
readline = sys.stdin.readline

N = int(readline())
T, P = [], []
for _ in range(N):
    t, p = map(int, readline().split())
    T.append(t)
    P.append(p)

d = [0] * (N+1)

for i in range(N - 1, -1, -1):
    if i + T[i] > N:
        d[i] = d[i+1]
    else:
        d[i] = max(d[i+1], P[i] + d[i + T[i]])

print(d[0])

File: 백준/Silver/14501. 퇴사/퇴사.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
let N = 0;
let arr = [];
rl.on("line", (line) => {
  if (inputLine == 0) {
    N = parseInt(line);
  }
  if (inputLine > 0 && inputLine <= N) {
    arr.push(line.split(" ").map((el) => parseInt(el)));
  }
  if (inputLine == N) {
    let dp = new Array(N).fill(0);
    for (let i = 0; i < N; i++) {
      let [p, t] = arr[i];
      if (i + p > N) continue;
      dp[i] = dp[i] + t;
      for (let j = i + p; j < N; j++) {
        dp[j] = Math.max(dp[j], dp[i]);
      }
    }
    console.log(Math.max(...dp));
    process.exit();
  }
  inputLine++;
});
